### PR TITLE
Implement Console.SetWindowSize() for linux

### PR DESCRIPTION
**Pull request by HJLeee**
Implement `Console.SetWindowSize()` using the same system call with `ConsoleGetWindowSize()` (`ioctl` w/ set param TIOCSWINSZ).

### CODE REVIEW

File: src/libraries/System.Console/tests/WindowAndCursorProps.cs
Comment (MichalPetryka) [inline src/libraries/System.Console/tests/WindowAndCursorProps.cs:107]: Shouldn't those tests be left here and just changed to only Browser, iOS, MacCatalyst and tvOS?

File: src/libraries/System.Console/src/System/ConsolePal.Unix.cs
Comment (stephentoub) [inline src/libraries/System.Console/src/System/ConsolePal.Unix.cs:406]: These should be moved to Console.SetWindowSize prior to delegating to ConsolePal.SetWindowSize, and then removed from both here and from ConsolePal.Windows.cs.

File: src/libraries/System.Console/src/System/ConsolePal.Unix.cs
Comment (stephentoub) [inline src/libraries/System.Console/src/System/ConsolePal.Unix.cs:421]: This should continue to throw a PlatformNotSupportedException if it's failing because Interop.Sys.SetWindowSize isn't implemented on the target platform.

File: src/libraries/System.Console/tests/WindowAndCursorProps.cs
Comment (stephentoub) [reply src/libraries/System.Console/tests/WindowAndCursorProps.cs:107]: Yes, please.

File: src/libraries/apicompat/ApiCompatBaseline.NetCoreAppLatestStable.txt
Comment (stephentoub) [inline src/libraries/apicompat/ApiCompatBaseline.NetCoreAppLatestStable.txt:143]: @ViktorHofer, is there a bug in the compat tooling, or is something wrong in the implementation in this PR?  This text says `changed from '[SupportedOSPlatformAttribute("windows")]' in the contract to '[SupportedOSPlatformAttribute("linux")]' in the implementation`, but shouldn't it say that it changed to be both windows and linux, not just linux?

File: src/libraries/System.Console/tests/WindowAndCursorProps.cs
Comment (HJLeee) [reply src/libraries/System.Console/tests/WindowAndCursorProps.cs:107]: Fixed. Thank you for the info. :)

File: src/libraries/System.Console/src/System/ConsolePal.Unix.cs
Comment (HJLeee) [reply src/libraries/System.Console/src/System/ConsolePal.Unix.cs]: Fixed. Thank you.

File: src/libraries/System.Console/src/System/ConsolePal.Unix.cs
Comment (HJLeee) [reply src/libraries/System.Console/src/System/ConsolePal.Unix.cs]: Fixed. Thank you.

File: src/libraries/System.Console/tests/WindowAndCursorProps.cs
Comment (MichalPetryka) [inline src/libraries/System.Console/tests/WindowAndCursorProps.cs:74]: Why is Browser supposed to not throw here but throw for WindowWidth?

Comment (HJLeee) [reply src/libraries/System.Console/tests/WindowAndCursorProps.cs:74]: `[PlatformSpecific(TestPlatforms.AnyUnix & ~TestPlatforms.Browser)]` was original test condition.

Comment (HJLeee) [reply src/libraries/System.Console/tests/WindowAndCursorProps.cs:74]: https://github.com/dotnet/runtime/commit/0de50030b876e9c77368e6144e53ed1b540a517c is the commit where original condition introduced.

File: src/libraries/System.Console/src/System/Console.cs
Comment (danmoseley) [inline src/libraries/System.Console/src/System/Console.cs:453]: This isn't inside the lock, so is it possible the console could get redirected between this check and the call to the PAL, and if so would that matter?

Comment (HJLeee) [reply src/libraries/System.Console/src/System/Console.cs:453]: Thank you for the review.
As I see it from ConsolePal.Unix.cs, `Console.IsOurputRedirected` doesn't need lock and assumes it is fixed value.
https://github.com/dotnet/runtime/blob/7dad0f36e56161e1a92fc6660b8386d054439ff0/src/libraries/System.Console/src/System/ConsolePal.Unix.cs#L202-L203
https://github.com/dotnet/runtime/blob/7dad0f36e56161e1a92fc6660b8386d054439ff0/src/libraries/System.Console/src/System/ConsolePal.Unix.cs#L216-L219
https://github.com/dotnet/runtime/blob/7dad0f36e56161e1a92fc6660b8386d054439ff0/src/libraries/System.Console/src/System/ConsolePal.Unix.cs#L229-L232
https://github.com/dotnet/runtime/blob/7dad0f36e56161e1a92fc6660b8386d054439ff0/src/libraries/System.Console/src/System/ConsolePal.Unix.cs#L237-L238

File: src/libraries/apicompat/ApiCompatBaseline.NetCoreAppLatestStable.txt
Comment (ViktorHofer) [reply src/libraries/apicompat/ApiCompatBaseline.NetCoreAppLatestStable.txt:143]: I defer to @smasher164 who just recently reimplemented this rule in the APICompat tooling. @smasher164 I think you noticed that the logging in the old APICompat is misleading. As we will soon switch to the new tooling, has this situation gotten better now?

Comment (HJLeee) [reply src/libraries/apicompat/ApiCompatBaseline.NetCoreAppLatestStable.txt:143]: @smasher164 Would you comment on this? Is this OK for this PR?

Comment (ViktorHofer) [reply src/libraries/apicompat/ApiCompatBaseline.NetCoreAppLatestStable.txt:143]: The existing tooling is a bit off when it comes to these messages but that shouldn't block this PR. We will get this cleaned-up when we switch to the new roslyn based APICompat tooling.

**Comment by HJLeee** [comment]
@stephentoub Would you review this please?

File: src/native/libs/System.Native/pal_console.c
Comment (adamsitnik) [inline src/native/libs/System.Native/pal_console.c:48]: > The [ioctl(2)](https://man7.org/linux/man-pages/man2/ioctl.2.html) system call returns 0 on success.  On error, it
       returns -1 and sets [errno](https://man7.org/linux/man-pages/man3/errno.3.html) to indicate the error.

nit: Calling it `error` here suggests that the sys-call does not set the error but returns it instead. You can just return the value returned by `ioctl` 
```suggestion
    return ioctl(STDOUT_FILENO, TIOCSWINSZ, windowSize);
```

File: src/libraries/System.Console/tests/WindowAndCursorProps.cs
Comment (adamsitnik) [inline src/libraries/System.Console/tests/WindowAndCursorProps.cs:590]: All the tests that you have modified verify that the method throws when it's not supported or the input is invalid.

Please enable `SetWindowSize_GetWindowSize_ReturnsExpected` test for configs which actually support the new API.

https://github.com/dotnet/runtime/blob/8d6b52f2b58bad7b1c48d98305fbcc9dc6f56921/src/libraries/System.Console/tests/WindowAndCursorProps.cs#L562-L586

File: src/libraries/System.Console/tests/WindowAndCursorProps.cs
Comment (adamsitnik) [inline src/libraries/System.Console/tests/WindowAndCursorProps.cs:114]: I can't see `OSX` and `FreeBSD` here, does it mean that the method is now supported on OSX and BSD? If so, you should add more `SupportedOSPlatform` attributes to the ref assembly.

File: src/libraries/System.Console/src/System/ConsolePal.Unix.cs
Comment (adamsitnik) [inline src/libraries/System.Console/src/System/ConsolePal.Unix.cs:352]: nit: you could use expression here
```suggestion
            set => SetWindowSize(value, WindowHeight);
```

**Review by adamsitnik — COMMENTED** [review]
@HJLeee thank you for your contribution! PTAL at my comments.

File: src/libraries/System.Console/src/System/ConsolePal.Unix.cs
Comment (HJLeee) [inline src/libraries/System.Console/src/System/ConsolePal.Unix.cs:365]:
```suggestion
            set => SetWindowSize(WindowWidth, value);
```

File: src/libraries/System.Console/tests/WindowAndCursorProps.cs
Comment (HJLeee) [reply src/libraries/System.Console/tests/WindowAndCursorProps.cs:114]: Thank you. `FreeBSD` and `OSX` are added to `SupportedOSPlatform`.

File: src/libraries/System.Console/tests/WindowAndCursorProps.cs
Comment (HJLeee) [reply src/libraries/System.Console/tests/WindowAndCursorProps.cs:590]: Thank you for the comment.
Regarding the positive test, I was wondering if `SetWindowSize_GetWindowSize_ReturnsExpected` actually tests `Console.SetWindowSize` on Windows test env.
On linux test env, `Console.IsOutputRedirected` is `true` and it makes whole test context skipped.
I can confirm that it works on linux when console output is not redirected.

Comment (HJLeee) [reply src/libraries/System.Console/tests/WindowAndCursorProps.cs:590]: [5150fcc](https://github.com/dotnet/runtime/pull/75824/commits/5150fcc33970c26caab2731e2afa36d40a397297) is pushed for consistency with Windows

Comment (HJLeee) [reply src/libraries/System.Console/tests/WindowAndCursorProps.cs:590]: @adamsitnik Please review this.

**Review by adamsitnik — COMMENTED** [review]
HI @HJLeee 

Please excuse me for the delay in the review.

I've cloned your fork and tried to add some new [console manual tests](https://github.com/dotnet/runtime/tree/main/src/libraries/System.Console/tests/ManualTests):

```cs
        [ConditionalFact(nameof(ManualTestsEnabled))]
        [SkipOnPlatform(TestPlatforms.Browser | TestPlatforms.iOS | TestPlatforms.MacCatalyst | TestPlatforms.tvOS, "Not supported on Browser, iOS, MacCatalyst, or tvOS.")]
        public static void ResizeTest()
        {
            bool wasResized = false;

            using (ManualResetEvent manualResetEvent = new(false))
            using (PosixSignalRegistration.Create(PosixSignal.SIGWINCH,
               ctx =>
               {
                   wasResized = true;
                   Assert.Equal(PosixSignal.SIGWINCH, ctx.Signal);
                   manualResetEvent.Set();
               }))
            {
                int widthBefore = Console.WindowWidth;
                int heightBefore = Console.WindowHeight;

                Assert.False(wasResized);

                Console.SetWindowSize(widthBefore / 2, heightBefore / 2);

                Assert.True(manualResetEvent.WaitOne(TimeSpan.FromMilliseconds(50)));
                Assert.True(wasResized);
                Assert.Equal(widthBefore / 2, Console.WindowWidth );
                Assert.Equal(heightBefore / 2, Console.WindowHeight );

                try
                {
                    AssertUserExpectedResults("terminal window getting two times smaller?");
                }
                finally
                {
                    Console.SetWindowSize(widthBefore, heightBefore);
                }
            }
        }
```

The problem is that the value is set correctly (and we get signal notification), but Terminal size does not change in visible way (I've tested GNOME and xterm terminals on Ubuntu). According to [this](https://stackoverflow.com/a/11491315) StackOverflow answer it does not work by design on most of the terminals and we should use escape sequences to get it working. We are already using some: 

https://github.com/dotnet/runtime/blob/7fbf1c134040f8cf7f6a807d14bdaa323d109456/src/libraries/System.Console/src/System/ConsolePal.Unix.cs#L834-L840

Which Terminal have you used for testing?

To run any dotnet app using your local build of dotnet/runtime you need to find corerun:

```cmd
find -name corerun
```

and run it by providing the path to the dll:

```cmd
./artifacts/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/8.0.0/corerun $pathTo.dll
```

**Comment by HJLeee** [comment]
Hi, Thank you for the testcase and review.
>Which Terminal have you used for testing?

I've used my ubuntu 22.04 with `screen-256color` TERM and checked effectiveness with `nvim` as below.
You can see `nvim` uses the value I set for testing.
```csharp
· ~/dotnet/tests/winsize $ cat Program.cs
// See https://aka.ms/new-console-template for more information
Console.SetWindowSize(80,30);
Console.WriteLine("Width: {0}, Height: {1}", Console.WindowWidth, Console.WindowHeight);
```
캡처

**Comment by HJLeee** [comment]
Rebased and force pushed to resolve conflict.

File: src/libraries/System.Console/src/System/Console.cs
Comment (danmoseley) [inline src/libraries/System.Console/src/System/Console.cs:414]: aside- we can delete
https://github.com/dotnet/runtime/blob/399c6dc66ecf89bda90a6070da194fcc639e15ef/src/libraries/System.Private.CoreLib/src/System/ArgumentOutOfRangeException.cs#L51-L53
https://github.com/dotnet/runtime/blob/399c6dc66ecf89bda90a6070da194fcc639e15ef/src/libraries/System.Private.CoreLib/src/System/ArgumentOutOfRangeException.cs#L90-L92

I dont know why we wouldn't use it in the libraries, as you are doing here, and indeed I see we are already doing in some places eg
https://github.com/dotnet/runtime/blob/399c6dc66ecf89bda90a6070da194fcc639e15ef/src/libraries/Common/src/System/IO/StreamHelpers.CopyValidation.cs#L19

It makes a nicer message.

Comment (HJLeee) [reply src/libraries/System.Console/src/System/Console.cs:414]: Thank you for the note.
These are where I've referred to
https://github.com/dotnet/runtime/blob/60be1deb89d1d811bd0380056d7831ba622d9409/src/libraries/System.Console/src/System/Console.cs#L547-L551
https://github.com/dotnet/runtime/blob/60be1deb89d1d811bd0380056d7831ba622d9409/src/libraries/System.Console/src/System/ConsolePal.Windows.cs#L672-L686

File: src/libraries/System.Console/tests/ManualTests/ManualTests.cs
Comment (adamsitnik) [inline src/libraries/System.Console/tests/ManualTests/ManualTests.cs:360]: Please excuse me, my assumptions about the visible effects of this action were wrong.
```suggestion
                Console.SetWindowSize(widthBefore, heightBefore);
```

File: src/libraries/System.Console/ref/System.Console.cs
Comment (adamsitnik) [inline src/libraries/System.Console/ref/System.Console.cs:72]: @buyaa-n @stephentoub So far, getting windows height and width was working on all OSes except browser, ios, adnroid and tvos.
Setting the value was working only on Windows. So the attributes applied were like this:

```cs
public static int WindowHeight
{ 
    [UnsupportedOSPlatformAttribute("android")]
    [UnsupportedOSPlatformAttribute("browser")]
    [UnsupportedOSPlatformAttribute("ios")]
    [UnsupportedOSPlatformAttribute("tvos")]
    get { throw null; } 
    [SupportedOSPlatformAttribute("windows")]
    set { } }
}
```

With this PR, both sets of supported/unsupported OSes are the same:
- supported: Windows, Linux, macOS, FreeBSD
- unsupported: Android, iOS, tvOS, Broweser

The question: what is the recommended way to annotate the APIs in such case:
a) marking getter and setter of WindowHeight as unsupported on the mentioned platforms (a removal of `[SupportedOSPlatformAttribute("windows")]` from the setter)

```cs
[UnsupportedOSPlatformAttribute("android")]
[UnsupportedOSPlatformAttribute("browser")]
[UnsupportedOSPlatformAttribute("ios")]
[UnsupportedOSPlatformAttribute("tvos")]
public static int WindowHeight
{ 

    get { throw null; } 
    set { } }
}
```

b) Extending the list of supported OSes for the setter (no breaking changes, but using two different ways of expressing the same thing for one public property which can be confusing)

```cs
public static int WindowHeight
{ 
    [UnsupportedOSPlatformAttribute("android")]
    [UnsupportedOSPlatformAttribute("browser")]
    [UnsupportedOSPlatformAttribute("ios")]
    [UnsupportedOSPlatformAttribute("tvos")]
    get { throw null; } 
    [SupportedOSPlatformAttribute("windows")]
    [SupportedOSPlatformAttribute("linux")]
    [SupportedOSPlatformAttribute("freebsd")]
    [SupportedOSPlatformAttribute("osx")]
    set { } }
}
```

**Review by adamsitnik — COMMENTED** [review]
Overall it looks good to me, I just need to clarify the recommended approach to attributes. Thank you @HJLeee !!

File: src/libraries/System.Console/ref/System.Console.cs
Comment (buyaa-n) [reply src/libraries/System.Console/ref/System.Console.cs:72]:
```cs
[UnsupportedOSPlatformAttribute("android")]
[UnsupportedOSPlatformAttribute("browser")]
[UnsupportedOSPlatformAttribute("ios")]
[UnsupportedOSPlatformAttribute("tvos")]
```
Looks cleaner though it means the API is unsupported on android, browser, ios and tvOS, but supported all other, not only windows, linux, freebsd, osx.  Not sure if that is OK.

For the PCA analyzer it doesn't really matter where the attribute applied, but not sure if it's moving the attributes is OK for API compat side

File: src/libraries/System.Console/src/System/Console.cs
Comment (stephentoub) [inline src/libraries/System.Console/src/System/Console.cs:411]: Nit:
```suggestion
                }

```

File: src/libraries/System.Console/src/System/Console.cs
Comment (stephentoub) [inline src/libraries/System.Console/src/System/Console.cs:415]: Nit:
```suggestion
                }

```

File: src/libraries/System.Console/src/System/Console.cs
Comment (stephentoub) [inline src/libraries/System.Console/src/System/Console.cs:432]: Ditto for the blank lines here

File: src/libraries/System.Console/src/System/Console.cs
Comment (stephentoub) [inline src/libraries/System.Console/src/System/Console.cs:457]: Ditto here, too

File: src/libraries/System.Console/src/System/ConsolePal.Unix.cs
Comment (stephentoub) [inline src/libraries/System.Console/src/System/ConsolePal.Unix.cs:415]: Nit: this could be consolidated a bit to:
```suggestion
                   throw errorInfo.Error == Interop.Error.ENOTSUP ?
                       new PlatformNotSupportedException() :
                       Interop.GetIOException(errorInfo);
```

**Review by stephentoub — APPROVED** [review]
LGTM. Thanks!

File: src/libraries/System.Console/ref/System.Console.cs
Comment (stephentoub) [reply src/libraries/System.Console/ref/System.Console.cs:72]: Is there a downside to option (a)?  That seems like the clear winner :)

File: src/libraries/System.Console/src/System/Console.cs
Comment (HJLeee) [inline src/libraries/System.Console/src/System/Console.cs:436]:
```suggestion
                }

```

File: src/libraries/System.Console/src/System/Console.cs
Comment (HJLeee) [inline src/libraries/System.Console/src/System/Console.cs:440]:
```suggestion
                }

```

File: src/libraries/System.Console/src/System/Console.cs
Comment (HJLeee) [inline src/libraries/System.Console/src/System/Console.cs:460]:
```suggestion
            }

```

File: src/libraries/System.Console/src/System/Console.cs
Comment (HJLeee) [inline src/libraries/System.Console/src/System/Console.cs:464]:
```suggestion
            }

```

File: src/libraries/System.Console/src/System/Console.cs
Comment (HJLeee) [inline src/libraries/System.Console/src/System/Console.cs:468]:
```suggestion
            }

```

File: src/libraries/System.Console/ref/System.Console.cs
Comment (HJLeee) [reply src/libraries/System.Console/ref/System.Console.cs]: @adamsitnik @buyaa-n I think I need help on https://github.com/dotnet/runtime/pull/75824/commits/e315d16aae366e45b9a416664ed54b2f5c36840f
How do I fix `CP0014` error?

**Review by adamsitnik — APPROVED** [review]
LGTM, big thanks for your contribution @HJLeee !

**Comment by adamsitnik** [comment]
> src/libraries/apicompat/ApiCompat.proj(50,5): error CP0014: (NETCORE_ENGINEERING_TELEMETRY=Build) Cannot remove attribute 'System.Runtime.Versioning.UnsupportedOSPlatformAttribute("android")' from 'System.Console.WindowHeight.get'.

@ViktorHofer we want to introduce this change on purpose, what is needed to make the CI green in such case?

**Comment by ViktorHofer** [comment]
Replied offline:

> You will need to add the compatibility difference to the suppression file that is used when comparing net7.0 against net6.0: [runtime/ApiCompatBaseline.NetCoreAppLatestStable.xml at main · dotnet/runtime (github.com)](https://github.com/dotnet/runtime/blob/main/src/libraries/apicompat/ApiCompatBaseline.NetCoreAppLatestStable.xml).
You can either hand-edit it or use the following command to update it: dotnet.cmd src/libraries/apicompat --no-dependencies /p:GenerateCompatibilitySuppressionFile=true

**Comment by adamsitnik** [comment]
@ViktorHofer I've done that and pushed the changes. I can perform a full clean build now (`build -c Release -subset clr+libs+libs.tests`) on both Windows and Linux, but the CI is still failing with the old error message (before the fix I was getting the same error locally). Do you have any idea why?

**Comment by ViktorHofer** [comment]
> @ViktorHofer I've done that and pushed the changes. I can perform a full clean build now (build -c Release -subset clr+libs+libs.tests) on both Windows and Linux, but the CI is still failing with the old error message (before the fix I was getting the same error locally). Do you have any idea why?

Looking now

**Comment by ViktorHofer** [comment]
The build is passing for me locally as well and I diffed the binlogs from the failing run and the local one. There aren't any differences that would explain this behavior.

After spending some time on it, I just noticed that CI wasn't triggered for your commit :(

**Comment by HJLeee** [comment]
Wow. Thank you for `APICompat` and CI stuff. 😄

**Pull request merged**
